File: 02-week-2--recursion/03-day-3--recursive-fibonacci-series/javascript/fibonacci_recursive.js
function fibonacci(n) {
  // type your code here
  let obj = {
    "0": 0,
    "1": 1
  }

  let i = 1

  const incrementObj = (index) => {
    const previousIndex = (index - 1)
    const nextIndex = (index + 1)
    obj[`${nextIndex}`] = obj[`${previousIndex}`] + obj[`${index}`]
    return nextIndex
  }

  const populateObj = () => {
      i = incrementObj(i)
      if (!!obj[`${n}`] || i == n + 3) {
        return obj[`${n}`]
      } else {
        return populateObj()
      }
  }

  return populateObj()
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log("=>", fibonacci(0));

  console.log("");

  console.log("Expecting: 1");
  console.log("=>", fibonacci(2));

  console.log("");

  console.log("Expecting: 55");
  console.log("=>", fibonacci(10));
}

module.exports = fibonacci;

// Please add your pseudocode to this file
/*
  -create hash with keys/values representing first two numbers of fibonacci series
  -create an incremental function which adds keys to this hash
  -create another function which calls the incremental one and checks to see if the hash has the desired key yet
  -if the hash has the desired key return the desired key, else the function calls itself again.
*/
// And a written explanation of your solution
/*
  -I decided creating a hash would make for a faster solve than iterating constantly through an array. so first thing is I
  created a hash with keys for the first two index values for the fibonacci series.  I then made a function that incremented
  a variable representing the current and used this to add keys and values to this hash.  I then made another function which
  was recursive and would call this incrementing function, see if the hash had the desired key yet, if so return the desired
  key, if not call itself again.  I returned this function at the end of the  
*/

File: 02-week-2--recursion/03-day-3--recursive-fibonacci-series/ruby/fibonacci_recursive.rb
def fibonacci(n)
  # type your code in here
  if n == 0 || n == 1
    return n
  end
  fibonacci(n - 1) + fibonacci(n - 2)
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts "=>", fibonacci(0)

  puts

  puts "Expecting: 1"
  puts "=>", fibonacci(2)

  puts

  puts "Expecting: 55"
  puts "=>", fibonacci(10)

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
=begin
  -if n < 2 return n
  -else decrease n by 1 and add it to n decreased by 2 until condition is met, calling the function recursively
=end
# And a written explanation of your solution
=begin
  Because I was only able to hold onto one number throughout this I needed to turn the problem into a series of 1 + 1 + 1...etc.
  The only way to store all of these ones was to make a condition at the start of the method that would return n if it were ever one
  However I also needed to account for if n was originally 0 so 0 needed to be returned if this was the case, hence the or statement.
  If I added two calls of this method together while decreasing n each time they were called I would eventually get the series of
  addition (1 + 1 + 1...etc) that I had hoped for.  Once all of these method calls had values the return would add them all together.
=end
